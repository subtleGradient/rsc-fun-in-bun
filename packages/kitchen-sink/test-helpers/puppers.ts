import packageJson from "#package"
import { afterAll, beforeAll } from "bun:test"
import puppeteer, { Browser, ConsoleMessage, Page } from "puppeteer-core"

let browser: Browser

////////////////////////////////////////////////////////////////
// duck punching shennanigans

declare module "puppeteer-core" {
  interface Browser {
    [Symbol.asyncDispose](): Promise<void>
  }
  interface Page {
    [Symbol.asyncDispose](): Promise<void>
    pipeConsoleLogs: boolean
    $pipeConsoleLogs_isEnabled: boolean
  }
}

Browser.prototype[Symbol.asyncDispose] = async function () {
  try {
    if (this.connected) await this.close()
  } catch (e) {
    console.warn("Browser.prototype[Symbol.asyncDispose]", e)
  }
}

Page.prototype[Symbol.asyncDispose] = async function () {
  try {
    if (!this.isClosed()) await this.close({ runBeforeUnload: false })
  } catch (e) {
    console.warn("Page.prototype[Symbol.asyncDispose]", e)
  }
}

Object.defineProperties(Page.prototype, {
  pipeConsoleLogs_isEnabled: { configurable: true, enumerable: false, writable: true, value: false },
  pipeConsoleLogs: {
    configurable: true,
    enumerable: true,
    get(this: Page) { return this.$pipeConsoleLogs_isEnabled }, // prettier-ignore
    set(this: Page, isEnabled: boolean) {
      if ((this.$pipeConsoleLogs_isEnabled = isEnabled)) {
        this.on("console", async (msg: ConsoleMessage) => {
          const args = await Promise.all(msg.args().map(it => it.toString()))
          // @ts-ignore -- trust me bro ðŸ˜Ž
          console[msg.type()](...args)
        })
        this.on("error", console.error)
      } else {
        this.off("console")
        this.off("error")
      }
    },
  },
})

////////////////////////////////////////////////////////////////

beforeAll(async () => {
  const headless = JSON.parse(process.env.PUPPETEER_HEADLESS ?? "false") as boolean
  if (!headless) console.warn("set PUPPETEER_HEADLESS=true to run headless")

  const executablePath =
    process.env.PUPPETEER_EXECUTABLE_PATH ?? "/Applications/Google Chrome.app/Contents/MacOS/Google Chrome"

  console.assert(
    await Bun.file(executablePath).exists(),
    "Chrome executable not found, set PUPPETEER_EXECUTABLE_PATH env var",
  )

  browser = await puppeteer.launch({
    executablePath,
    headless,
    defaultViewport: null,
    waitForInitialPage: false,
    dumpio: true,
    args: [
      "--no-startup-window", // each new page will open in a new window instead of tabs
      `--user-data-dir=${process.env.TMPDIR ?? "/tmp/"}/puppers-${packageJson.name.replace(/[^a-z0-9_.$]/gi, "-")}`,
    ],
  })

  process.once("beforeExit", async () => await browser.close())
})

afterAll(() => browser.close())

export { browser, puppeteer }
