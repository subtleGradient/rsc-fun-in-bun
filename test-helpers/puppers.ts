import { afterAll, beforeAll } from "bun:test"
import puppeteer, { Browser } from "puppeteer-core"

let browser: Browser

beforeAll(async () => {
  const headless = JSON.parse(process.env.PUPPETEER_HEADLESS ?? "false") as boolean
  if (!headless) console.warn("set PUPPETEER_HEADLESS=true to run headless")

  const executablePath =
    process.env.PUPPETEER_EXECUTABLE_PATH ?? "/Applications/Google Chrome.app/Contents/MacOS/Google Chrome"

  console.assert(
    await Bun.file(executablePath).exists(),
    "Chrome executable not found, set PUPPETEER_EXECUTABLE_PATH env var",
  )

  browser = await puppeteer.launch({
    executablePath,
    headless,
    defaultViewport: null,
    waitForInitialPage: false,
    args: [
      "--no-startup-window", // each new page will open in a new window instead of tabs
    ],
  })
})

afterAll(() => browser.close())

// return Object.assign(browser, { [Symbol.dispose]: () => browser.close() })

export { browser, puppeteer }
