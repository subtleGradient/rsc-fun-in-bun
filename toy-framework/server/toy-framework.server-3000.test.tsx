import { browser } from "@/test-helpers/puppers.ts"
import { describe, expect, it } from "bun:test"

const RSC_TYPE = "text/x-component"

if (!1!) {
  // trigger the tests to re-run when this file changes
  // const toyFramework = await import("./toy-framework.server.tsx")
  const toyFramework = require("./toy-framework.server.tsx")

  describe("the `await using` keyword", async () => {
    it("disposes of the page auto-magically", async () => {
      await using page = await browser.newPage()
      await page.goto("about:blank")
      // no need to call page.close() here! ðŸ¤¯
    })
  })
}

describe("toy-framework.server", () => {
  describe("root / route", () => {
    it("returns a response", async () => {
      const response = await fetch(new Request(`http://localhost:3000`))
      expect(response.status).toBe(200)
      expect((await response.text()).replaceAll("><", ">\n<")).toMatchSnapshot()
    })

    it("renders ClientComponent", async () => {
      await using page = await browser.newPage()
      page.goto(`http://localhost:3000`)
      await page.waitForSelector("#generated-by-client")
      expect(await page.$eval("#generated-by-client", el => el.textContent)).toBe("Hello from ClientComponent!")
    })
  })

  describe("/rsc/test-suspense", () => {
    it("returns RSC content", async () => {
      const response = await fetch(new Request(`http://localhost:3000/rsc/test-suspense`, { headers: { Accept: RSC_TYPE } }))
      expect(response.status).toBe(200)
      expect(await response.text()).toMatchSnapshot()
    })
    it("updates asynchronously", async done => {
      const rscResponse = await fetch(
        new Request(`http://localhost:3000/rsc/test-suspense`, { headers: { Accept: RSC_TYPE } }),
      )
        // stream the responce and verify that it sends multiple chunks
      const decoder = new TextDecoder()
      const reader = rscResponse.body!.getReader()
      reader.closed.then(done)
      let count = 0
      while (true) {
        const { done, value } = await reader.read()
        if (done) break
          // expect(decoder.decode(value)).toMatchSnapshot()
        count++
      }
      expect(count).toBeGreaterThan(1)
      },
      { retry: 3 },
    )
  })

  describe("/rsc/test-suspense-render", () => {
    it("returns RSC content", async () => {
      const response = await fetch(new Request(`http://localhost:3000/rsc/test-suspense-render`))
      expect(response.status).toBe(200)
      expect(await response.text().then(text => text.replaceAll("><", ">\n<"))).toMatchSnapshot()
    })

    it("renders ClientComponent", async () => {
      await using page = await browser.newPage()
      page.goto(`http://localhost:3000/rsc/test-suspense-render`)
      await page.waitForSelector("#AsyncView10")
      expect(await page.$eval("#AsyncView10", el => el.textContent)).toBe("slept for 10ms")
    })
  })
})
