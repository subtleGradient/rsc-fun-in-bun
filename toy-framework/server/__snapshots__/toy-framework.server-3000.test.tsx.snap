// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`toy-framework.server root / route returns a response 1`] = `
"<!DOCTYPE HTML>
<HTML lang=en>
<script type="importmap">{
  "imports": {
    "bun": "/!/externals:bun.mjs",
    "react": "/!/externals/react.mjs",
    "react-dom": "/!/externals/react-dom.mjs",
    "react-dom/client": "/!/externals/react-dom:client.mjs",
    "react-server-dom-webpack/client": "/!/externals/react-server-dom-webpack:client.mjs",
    "react/jsx-dev-runtime": "/!/externals/react:jsx-dev-runtime.mjs",
    "react/jsx-runtime": "/!/externals/react:jsx-runtime.mjs"
  }
}</script>
<head>
<title>Hello from /HTMLPageRootLayout.tsx</title>
<link rel="modulepreload" href="/!/polyfillsAndStuff.mjs"/>
<link rel="modulepreload" href="/!/externals/clientEntryPoint.externals.mjs"/>
<link rel="modulepreload" href="/!/externals/react.mjs"/>
<link rel="modulepreload" href="/!/externals/react-dom.mjs"/>
<link rel="modulepreload" href="/!/externals/react-dom:client.mjs"/>
<link rel="modulepreload" href="/!/externals/react-server-dom-webpack:client.mjs"/>
<link rel="modulepreload" href="/!/externals/react:jsx-dev-runtime.mjs"/>
<link rel="modulepreload" href="/!/externals/react:jsx-runtime.mjs"/>
<link rel="modulepreload" href="/!/clientEntryPoint/clientEntryPoint.mjs"/>
<link rel="modulepreload" href="/!/use-client/example0.client.mjs"/>
</head>
<body>
<h1>Hello from <!-- -->/HTMLPageRootLayout.tsx</h1>
<div id="root">
<div data-id="NOT-generated-by-client">pre-rendered by server in <code>/HTMLPageRootLayout.tsx</code>
</div>
</div>
<script type="module" src="/!/polyfillsAndStuff.mjs">
</script>
<script type="module" src="/!/externals/clientEntryPoint.externals.mjs">
</script>
<script type="module" src="/!/clientEntryPoint/clientEntryPoint.mjs">
</script>
</body>
</HTML>"
`;

exports[`toy-framework.server /rsc/test-suspense returns RSC content 1`] = `
"1:"$Sreact.suspense"
2:D{"name":"AsyncView","env":"Server"}
3:D{"name":"AsyncView","env":"Server"}
0:["$","div",null,{"children":["hi",["$","$1",null,{"fallback":["$","div",null,{"children":"loading..."}],"children":["$L2","$L3"]}]]}]
2:["$","div",null,{"id":"AsyncView0","children":["slept for ",0,"ms"]}]
3:["$","div",null,{"id":"AsyncView10","children":["slept for ",10,"ms"]}]
"
`;

exports[`toy-framework.server /rsc/test-suspense-render returns RSC content 1`] = `
"<!DOCTYPE HTML>
<HTML lang=en>
<script type="importmap">{
  "imports": {
    "bun": "/!/externals:bun.mjs",
    "react": "/!/externals/react.mjs",
    "react-dom": "/!/externals/react-dom.mjs",
    "react-dom/client": "/!/externals/react-dom:client.mjs",
    "react-server-dom-webpack/client": "/!/externals/react-server-dom-webpack:client.mjs",
    "react/jsx-dev-runtime": "/!/externals/react:jsx-dev-runtime.mjs",
    "react/jsx-runtime": "/!/externals/react:jsx-runtime.mjs"
  }
}</script>
<head>
<title>Hello from /HTMLPageRootLayout.tsx</title>
<link rel="modulepreload" href="/!/polyfillsAndStuff.mjs"/>
<link rel="modulepreload" href="/!/externals/clientEntryPoint.externals.mjs"/>
<link rel="modulepreload" href="/!/externals/react.mjs"/>
<link rel="modulepreload" href="/!/externals/react-dom.mjs"/>
<link rel="modulepreload" href="/!/externals/react-dom:client.mjs"/>
<link rel="modulepreload" href="/!/externals/react-server-dom-webpack:client.mjs"/>
<link rel="modulepreload" href="/!/externals/react:jsx-dev-runtime.mjs"/>
<link rel="modulepreload" href="/!/externals/react:jsx-runtime.mjs"/>
<link rel="modulepreload" href="/!/clientEntryPoint/clientEntryPoint.mjs"/>
<link rel="modulepreload" href="/!/use-client/example0.client.mjs"/>
</head>
<body>
<h1>Hello from <!-- -->/HTMLPageRootLayout.tsx</h1>
<div id="root">
<div data-id="NOT-generated-by-client">pre-rendered by server in <code>/HTMLPageRootLayout.tsx</code>
</div>
</div>
<script type="module" src="/!/polyfillsAndStuff.mjs">
</script>
<script type="module" src="/!/externals/clientEntryPoint.externals.mjs">
</script>
<script type="module" src="/!/clientEntryPoint/clientEntryPoint.mjs">
</script>
<h2>Howdy!</h2>
<h3>testing RSC</h3>
<div id="rsc-root">
</div>
<script type="module">;(async function main() {
    console.log("running", import.meta.url);
    const React = (await import("react")).default;
    console.log("react@" + React.version);
    const ReactDOMClient = (await import("react-dom/client")).default;
    const ReactServerDOMClient = (await import("react-server-dom-webpack/client")).default;
    console.log(ReactServerDOMClient);
    const rscResponse = await fetch(new Request(\`/rsc/test-suspense\`, { headers: { Accept: "text/x-component" } }));
    const decoder = new TextDecoder;
    const reader = rscResponse.body.getReader();
    while (true) {
      const { done, value } = await reader.read();
      if (done)
        break;
      console.log(decoder.decode(value));
    }
    const root = ReactDOMClient.createRoot(document.getElementById("rsc-root"));
    const ui = await ReactServerDOMClient.createFromFetch(fetch(new Request(\`/rsc/test-suspense\`, { headers: { Accept: "text/x-component" } })), {
      async callServer(id, args) {
        throw new Error("not implemented");
      },
      encodeFormAction(id, args) {
        return {
          name: "my-awesome-form",
          action: "/rsc/test-suspense",
          method: "POST",
          data: new FormData
        };
      }
    });
    root.render(ui);
  })();</script>
</body>
</HTML>"
`;
